cirq/circuits/circuit_gate.py

```diff
@@ -0,0 +1,190 @@
+# Copyright 2020 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""A structure for encapsulating entire circuits in a gate.
+
+A CircuitGate is a Gate object that wraps a FrozenCircuit. When applied as part
+of a larger circuit, a CircuitGate will execute all component gates in order,
+including any nested CircuitGates.
+"""
+
+from typing import TYPE_CHECKING, Any, Dict, Optional, Sequence, Tuple, Union
+
+from cirq import circuits, devices, ops, protocols
+from cirq.circuits.insert_strategy import InsertStrategy
+from cirq.type_workarounds import NotImplementedType
+
+if TYPE_CHECKING:
+    import cirq
+
+
+class CircuitGate(ops.Gate):
+    """A single gate that encapsulates a Circuit."""
+
+    def __init__(self,
+                 *contents: 'cirq.OP_TREE',
+                 strategy: 'cirq.InsertStrategy' = InsertStrategy.EARLIEST,
+                 device: 'cirq.Device' = devices.UNCONSTRAINED_DEVICE,
+                 name: Optional[str] = None,
+                 exp_modulus: Optional[int] = None) -> None:
```

Comment:
I'm not sure this is a good idea. When we ask hardware to repeat a circuit we typically want it to run that many times, we don't want things that are equivalent to identities to get optimized away (think about randomized benchmarking). Such transformations could be implemented in an optimization pass, but it seems odd to bake this in to the gate itself.